File: frontend/src/engine/capabilities/writeValues.ts
/**
 * writeValues – Write a 2D array of values to a range.
 *
 * CRITICAL: By default, this writes ONLY values. It does NOT copy or alter
 * any formatting (fill, font, borders, number formats). This is the primary
 * mechanism for preserving existing workbook formatting.
 *
 * Office.js notes:
 * - Setting range.values writes raw values; formulas in the values array
 *   will be written as literal strings unless they start with "=".
 * - The values array dimensions must match the target range dimensions.
 * - Null values in the array preserve the existing cell value.
 */

import { CapabilityMeta, WriteValuesParams, StepResult, ExecutionOptions } from "../types";
import { registry } from "../capabilityRegistry";
import { resolveRange } from "./rangeUtils";

const meta: CapabilityMeta = {
  action: "writeValues",
  description: "Write values to a cell range (preserves formatting)",
  mutates: true,
  affectsFormatting: false,
};

async function handler(
  context: Excel.RequestContext,
  params: WriteValuesParams,
  options: ExecutionOptions
): Promise<StepResult> {
  const { range: address, values, valuesOnly: _valuesOnly = true } = params;

  if (options.dryRun) {
    return {
      stepId: "",
      status: "success",
      message: `Would write ${values.length} rows to ${address}`,
    };
  }

  options.onProgress?.(`Writing ${values.length} rows to ${address}...`);

  // Resolve the range, then resize to match the actual values dimensions.
  // The LLM sometimes gets the range size wrong (e.g. C31:C400 for 366 rows).
  // We take just the top-left cell of the given range and resize from there.
  const fullRange = resolveRange(context, address);
  const topLeft = fullRange.getCell(0, 0);
  const rows = values.length;
  const cols = values[0]?.length ?? 1;
  const target = topLeft.getResizedRange(rows - 1, cols - 1);

  // Only set values — never touch formatting
  target.values = values;

  await context.sync();

  return {
    stepId: "",
    status: "success",
    message: `Wrote ${values.length} rows to ${address}`,
  };
}


registry.register(meta, handler as any);
export { meta };

File: backend/app/execution/capabilities/write_values.py
"""writeValues — bulk write a 2D array of values to a range."""

from __future__ import annotations

from typing import Any

from app.execution.base import ExecutorContext
from app.execution.capability_registry import registry
from app.execution.range_utils import resolve_range


def handler(ctx: ExecutorContext, params: dict[str, Any]) -> dict[str, Any]:
    address = params.get("range")
    values = params.get("values")
    if not address or values is None:
        return {"status": "error", "message": "writeValues requires 'range' and 'values'."}

    # Accept flat lists — reshape into 2D to match the frontend semantics.
    if isinstance(values, list) and (not values or not isinstance(values[0], list)):
        values = [values] if values else [[]]

    rng = resolve_range(ctx.workbook_handle, address)

    if ctx.dry_run:
        return {
            "status": "preview",
            "message": f"Would write {len(values)}×{len(values[0]) if values else 0} to {rng.address}.",
        }

    # Resize to match the incoming matrix so partial addresses like "A1"
    # auto-expand. Mirrors the frontend's `range.getResizedRange` behavior.
    rows = len(values)
    cols = max((len(r) for r in values), default=0)
    target = rng.resize(rows, cols) if hasattr(rng, "resize") else rng
    target.value = values

    return {
        "status": "success",
        "message": f"Wrote {rows}×{cols} to {target.address}.",
        "outputs": {"range": target.address, "rowCount": rows, "columnCount": cols},
    }


registry.register("writeValues", handler, mutates=True)
